modules/table/src/integrationTest/java/org/apache/ignite/internal/table/ItTxResourcesVacuumTest.java

```diff
@@ -0,0 +1,559 @@
+            + "  raft: { responseTimeout: 30000 },"
+            + "  compute.threadPoolSize: 1\n"
+            + "}";
+
+    private ExecutorService txStateStorageExecutor = Executors.newSingleThreadExecutor();
+
+    @BeforeEach
+    @Override
+    public void setup(TestInfo testInfo) throws Exception {
+        super.setup(testInfo);
+
+        String zoneSql = "create zone test_zone with partitions=20, replicas=" + REPLICAS
+                + ", storage_profiles='" + DEFAULT_STORAGE_PROFILE + "'";
+        String sql = "create table " + TABLE_NAME + " (key bigint primary key, val varchar(20)) with primary_zone='TEST_ZONE'";
+
+        cluster.doInSession(0, session -> {
+            executeUpdate(zoneSql, session);
+            executeUpdate(sql, session);
+        });
+
+        txStateStorageExecutor = Executors.newSingleThreadExecutor(IgniteThreadFactory.create("test", "tx-state-storage-test-pool", log,
+                ThreadOperation.STORAGE_READ));
+    }
+
+    @Override
+    @AfterEach
+    public void tearDown() {
+        shutdownAndAwaitTermination(txStateStorageExecutor, 10, TimeUnit.SECONDS);
+
+        super.tearDown();
+    }
+
+    @Override
+    protected void customizeInitParameters(InitParametersBuilder builder) {
+        super.customizeInitParameters(builder);
+
+        builder.clusterConfiguration("{"
+                + "  transaction: {"
+                + "      implicitTransactionTimeout: 30000,"
+                + "      txnResourceTtl: 2"
+                + "  },"
+                + "  replication: {"
+                + "      rpcTimeout: 30000"
+                + "  },"
+                + "}");
+    }
+
+    @Override
+    protected int initialNodes() {
+        return 3;
+    }
+
+    /**
+     * Returns node bootstrap config template.
+     *
+     * @return Node bootstrap config template.
+     */
+    @Override
+    protected String getNodeBootstrapConfigTemplate() {
+        return NODE_BOOTSTRAP_CFG_TEMPLATE;
+    }
+
+    @Test
+    public void testVacuum() throws InterruptedException {
+        IgniteImpl node = anyNode();
+
+        RecordView<Tuple> view = node.tables().table(TABLE_NAME).recordView();
+
+        // Put some value into the table.
+        Transaction tx = node.transactions().begin();
+        UUID txId = txId(tx);
+
+        log.info("Test: Loading the data [tx={}].", txId);
+
+        Tuple tuple = findTupleToBeHostedOnNode(node, TABLE_NAME, tx, INITIAL_TUPLE, NEXT_TUPLE, true);
+
+        int partId = partitionIdForTuple(node, TABLE_NAME, tuple, tx);
+
+        Set<String> nodes = partitionAssignment(node, new TablePartitionId(tableId(node, TABLE_NAME), partId));
+
+        view.upsert(tx, tuple);
+
+        tx.commit();
+
+        log.info("Test: Tx committed [tx={}].", txId);
+
+        checkVolatileTxStateOnNodes(nodes, txId);
+        waitForTxStateReplication(nodes, txId, partId, 10_000);
+
+        waitForTxStateVacuum(txId, partId, true, 10_000);
```

Comment:
In _every_ vacuum test I'd add check the data by reading corresponding expected values both with RW and RO transactions. It's also important to read the data with RO readTimestamp in between WI creation and vacuumization.
Mean following.
1. Tx1:: put(k1,v1), commit (commitTimestamp t10).
2. Tx2:: put(k1,v2), commit (commitTimestamp t20).
3. Vacuum both tx1, tx2 states.
4. RW T3:: get(k1) -> expect v2.
5. RO Tx4:: get(k1, /** readTimestamp */ t25) -> expect v2
6. RO Tx5:: get(k1, /** readTimestamp */ t15) -> expect v1
7. RO Tx6:: get(k1, /** readTimestamp */ t5) -> expect null

In order to "set" proper readTimestamp you can just start RO in proper moment in time and do enlist actual operation  whenever you actually need.

Reply:
In one test only?

Reply:
added more RO checks